ukify: add helper to create UKIs

keszybz opened the pull request:
```console
$ src/kernel-install/ukify.py --measure \
  /lib/modules/6.0.5-300.fc37.x86_64/vmlinuz \
  /boot/08a5690a2eed47cf92ac0a5d2e3cf6b0/6.0.5-300.fc37.x86_64/initrd \
  --key server.key --cert server.crt \
  --cmdline='rw quiet'

Host arch 'x86_64', efi arch 'x64'
+ objcopy /usr/lib/systemd/boot/efi/linuxx64.efi.stub --add-section .osrel=/etc/os-release --change-section-vma .osrel=0x20000 --add-section .cmdline=/tmp/tmpcmdline68txuxso --change-section-vma .cmdline=0x30000 --add-section .linux=/lib/modules/6.0.5-300.fc37.x86_64/vmlinuz --change-section-vma .linux=0x2000000 --add-section .initrd=/boot/08a5690a2eed47cf92ac0a5d2e3cf6b0/6.0.5-300.fc37.x86_64/initrd --change-section-vma .initrd=0x3000000 /tmp/linuxiofa0e0d
+ sbsign --key server.key --cert server.crt /tmp/linuxiofa0e0d --output vmlinuz.efi warning: data remaining[79844720 vs 79861844]: gaps between PE/COFF sections? warning: data remaining[79844720 vs 79861848]: gaps between PE/COFF sections? Signing Unsigned original image
Wrote signed vmlinuz.efi
+ /usr/lib/systemd/systemd-measure calculate --osrel /etc/os-release --cmdline /tmp/tmpcmdline68txuxso --linux /lib/modules/6.0.5-300.fc37.x86_64/vmlinuz --initrd /boot/08a5690a2eed47cf92ac0a5d2e3cf6b0/6.0.5-300.fc37.x86_64/initrd 11:sha1=55a84377aef93205f0d520410fd666db0cea842a
11:sha256=4eb94ecb4c7c7fca6dff9529d2efda20068b14e4eb5a36381345ac7d1051c5bb 11:sha384=9aee5234784fa59031d07f160349153a5b4b3ffaa4e33ba98e1906b9819a59b922e3220648e74217e5f6e5058cf982fa 11:sha512=84a7c9605c666886f12ba0519aad9232675d022b20969ef8904f6a0fdf9c3284fc965ea73aa561e202800596732b635dfa01fe93ca5e24e25265008ea9d1b6bc 11:sha1=b9a54ccafd8f2de90351859c9b41203186df2568
11:sha256=93ad91044a2f5491a31738cb50c5122b8e6109662cf7036362f47f892be859a9 11:sha384=6dcf39948bd567e09f7c38bd580e5acfd9f5f1a2e21b31f9b361b5ba920750ea9f46d3bb238a34cb86d7fd3397515ff4 11:sha512=bbb9856e96984a30df7ff12793fc2c6fa552a45417f07c40a534ebdd85f527768738bfcd19c149342d916b8863d985ca6170e9cc6208e7d11c6334fd79c5ece4 11:sha1=1d4c10141e53f1054c02f6d5fd31bd557b53f81d
11:sha256=df74c69e75067086f796a6947346638e7d2e06ba49ea6c464d0751e5f0695634 11:sha384=0a8c51511d3dcbce7378e45c82edba64944fc12435bd4060df86759b3b472c3de363f0c2c8ce6bbfacf136dc1ea44faf 11:sha512=01d4aeadd18c93e04e323a338be981a70836f5c60e12fbf2fff9f138ba529a6eabf15be9ce1e8af08b88f083f450f9a78e13a7f6cc2222c796823db27bfc2b3c 11:sha1=78ce3cf312a1b3bc4c19aac36158b30d6d6e74e3
11:sha256=69e63c7e6be0623113e259e249aeafb27ae232d122be978e74bf1a14df030153 11:sha384=20c39d7c0ea79bc2ff68e8c396926e2fdd4623dffa84d542f7652b5a98a369bd077f49f5b2217bbb8295daa7afe517f9 11:sha512=595c2bfca98c90b3baf614b2ec167c02bdedc6fda8084be88b8d6bd964cbfaea31fd6c7d738e0387b4001f42b57bec6652d253675a98464fb527227759ed1e22
```

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+
+        sections.append((name, content, offset))
+
+    add_section('osrel',   args.os_release,    0x20_000)
+    add_section('cmdline', args.cmdline,       0x30_000)
+    add_section('dtb',     args.devicetree,    0x40_000)
+    add_section('splash',  args.splash,       0x100_000)
+    add_section('linux',   args.linux,      0x2_000_000)
+    add_section('initrd',  args.initrd,     0x3_000_000)
```
It's a terrible practice that these values have been hard coded since forever. These really need to be picked based on actual size of the payload added and the largest VMA of the stub image. Especially if native PE image creation ever comes into the mix as that might generate large VMA offsets, meaning these fixed values make the final image too big to fit in memory.

Also, as long as there is fallback support for EFI handover protocol the linux section should really get some special attention:
 - it ought to be marked as executable with --set-section-flags
 - the decompressor expects enough free space after the kernel proper, so there needs to be breathing room after it from the PE header (SizeOfCode and/or SizeOfUninitializedData). The previous hard coded values were just lucky (or just chosen carefully). Really, I think the linux section should just go last. (Apparently, uncompressed aarch64 kernels can end up being too large for the default gap.)

medhefgo replied:
Here is why: https://github.com/systemd/systemd/issues/24202

And it wasn't just us hitting issues recently in CI, but people have been hitting this on older stubs as well (though, why people use that part of github to discuss things instead of filing a bug report is beyond me. it would've saved me so much time solving the above bug): https://github.com/systemd/systemd/commit/0fa2cac4f0cdefaf1addd7f1fe0fd8113db9360b#commitcomment-55421176

Foxboron replied:
The Arch Wiki was also updated with some bash to calculate the offsets.

https://wiki.archlinux.org/title/Unified_kernel_image#Manually

keszybz replied:
I tried to use`--set-section-flags`, but `objcopy` doesn't like that:

```console
$ objcopy /usr/lib/systemd/boot/efi/linuxx64.efi.stub --add-section .osrel=/etc/os-release --change-section-vma .osrel=0x4c38000 --add-section .cmdline=/tmp/tmpcmdline0p7h5fqp --change-section-vma .cmdline=0x4c39000 --add-section .pcrpkey=tpm2-pcr-public.pem --change-section-vma .pcrpkey=0x4c3a000 --add-section .initrd=/boot/08a5690a2eed47cf92ac0a5d2e3cf6b0/6.0.5-300.fc37.x86_64/initrd --change-section-vma .initrd=0x4c3b000 --add-section .uname=/tmp/tmpunameah3_h5y3 --change-section-vma .uname=0x8c1f000 --add-section .test=/tmp/tmptestp4jozmd4 --change-section-vma .test=0x8c20000 --add-section .test2=/tmp/tmptest252whqfze --change-section-vma .test2=0x8c21000 --add-section .pcrsig=/tmp/tmppcrsigllow37su --change-section-vma .pcrsig=0x8c22000 --add-section .linux=/lib/modules/6.0.5-300.fc37.x86_64/vmlinuz --change-section-vma .linux=0x8c23000 --set-section-flags .linux=contents,alloc,load,readonly,code /tmp/uki4f4pp6ao
objcopy: /tmp/uki4f4pp6ao: can't add section '.linux': file in wrong format
```

medhefgo replied:
What objcopy is this? It works for me on arch with 2.39.0. Afaik you only need to pass `--set-section-flags .linux=code,readonly`.

Ultimately, this isn't too important as is only really needed for the deprecated handover path and only out of paranoia. And stub assembly so far never set this as section as executable and nobody ever complained.

keszybz replied:
I guess I tried to pass too many flags. With `--set-section-flags .linux=code,readonly` it indeed works.

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
```
This should look at `/sys/firmware/efi/fw_platform_size` on x86.

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+    p.add_argument('--measure',
+                   action='store_true')
+
+    args = p.parse_args()
```
There should be a way to pass additional yet-unspecified sections too.

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+        output = args.output
+
+    objcopy = [
+        'objcopy',
```
Have you checked that this works for arched other than x86? Historically, not all arches have PE support in binutils (hence why those ship the elf stub too). And some tools have issues with the gnu-efi created PE binaries: if I remember correctly, sbsign doesn't like our arm binaries, and llvm-binutils definitely doesn't.

keszybz replied:
No, I haven't really tested this at all. Nevertheless, this essentially duplicates what dracut does and our docs say. I'm not even sure if I could test this properly. I have a rpi with 64 bits, but it has it's special bootloader IIRC, and wouldn't boot an UKI. Though we should definitely add some CI for this. Even if it turns out that we do the tool differently, it'd be useful to have tests. They can be adapted to a different scheme later.

bluca commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
```
build machine might be different than the target machine? there should be at least an option to manually override

medhefgo replied:
This is only looked at if `--efi-arch` wasn't passed.

keszybz replied:
The `--efi-arch` does this override. This function is only as a fallback.

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+        ('splash',  args.splash,       0x100_000),
+        ('initrd',  args.initrd,     0x3_000_000),
+        # linux shall be last to leave breathing room for decompression
+        ('linux',   args.linux,      0x2_000_000),
```
Reordering this here without giving it a different offset is not gonna do anything. And you cannot put it last without knowing the initrd size anyway.

Also, now that I think of this: it probably makes sense to put the linux section first and the osrel second section offset a suitable gap. That would guarantee there is a gap that does not contain anything important in it (all other sections are technically optional). If the linux section is last, the decompressor could conceivably override something important…

medhefgo replied:
> Why do we actually the sections like splash and dtb (which we wouldn't decompress)?

I can't parse this.

Basically, you look at the output of `objdump -h` or use the `pefile` python module to get at the data without ugly text parsing and then compute the offsets like this:

```
base_offset=highest_vma + highest_vma_size

if arch == x86 and kernel_PE_MajorImageVersion < 1:
    kernel_section_offset=align(base_offset, kernel_PE_OptionalHeader_SectionAlignment)
    kernel_size = MAX(kernel_file_size, kernel_PE_OptionalHeader_SizeOfCode)
else:
    kernel_section_offset=base_offset
    kernel_size = kernel_file_size

section1_offset=kernel_section_offset + kernel_size
section2_offset=section1_offset + section1_size
section3_offset=section2_offset + section2_size
…
```

medhefgo reviewed (commented):
I'd personally prefer if we made something like this the only officially supported way to assemble a UKI. This could even be pre-announced for the upcoming v252 to give other projects a heads up.

At the very least we need to add a warning to the changelog that hard-coded section values must not be used…

medhefgo commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
```
This should warn if the section name is longer than 8 chars. It'd be legal PE section name, but our stub cannot locate those. If this feature is used when prototyping new sections this can be caught early.

keszybz replied:
I added a check now. https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#section-table-section-headers says that anything utf-8 goes, but I think it's reasonable to limit ourselves to ascii now.

medhefgo commented:
Might also be a good idea to sign the payload kernel just in case.

behrmann commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+            unsigned.name,
+            '--output', args.output,
+        ]
+        print('+', shell_join(sbsign))
```
This one can be dismissed, as it is triggered by the `args.key` and isn't smart enough to understand it's just a path to a file. I wonder if CodeQL would not trigger if this were called `args.key_path` or similar.

behrmann commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+        subprocess.check_call(sbsign)
+
+        # We end up with no executable bits, let's reapply them
+        os.umask(umask := os.umask(0))
```
The walrus operator was introduced with Python 3.8. I don't know what the lower Python version requirement is, but if it's 3.8 you can use `shlex.join` as mentioned above.

behrmann commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
```
The above version is more in line with the `len(parts) >= 3` check in the offset parsing line. If you want to throw an error for `len(parts) > 3`, than that's not in this version and would need to be added. I think it's a bit more readable this way, though.

keszybz replied:
I think that an error should be thrown if there's too many parts. And then the version with `try..except` becomes noticably longer… Dunno, I think it's fine as is.

behrmann replied:
Yeah, I wasn't all too sure about this myself, because of the later check for `len(parts) >= 3`. It is fine as is, but adding another check does allow for nicer error messages, including @medhefgo's waning
```python       
try:
    name, content, *offset = s.split(':')
except ValueError as e:
    raise ValueError(f'Cannot parse section spec (name or content missing): {s!r}') from e
if len(offset) > 1:
    raise ValueError(f'Cannot parse section spec (more than one offset given): {s!r}')
if len(name) > 8:
    print('Warning: name of PE section is too long')
offset = int(offset[0], base=0) if offset else None

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
```
Unfortunately it's not so simple. Everything used here supports pathlib.Path transparently, but shlex.quote does not (in Python 3.11). Sadly.

behrmann replied:
Oh, wow, that seems like a pretty glaring omission on cpython's side. You're right.

behrmann replied:
The relevant issue there is python/cpython#89293

keszybz commented:
> Might also be a good idea to sign the payload kernel just in case.

I think that this should be done as a step of the building of the kernel, if at all.

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
+
+    if args.os_release is None:
+        p = pathlib.Path('/etc/os-release')
+        if not p.exists():
+            p = pathlib.Path('/usr/lib/os-release')
+        args.os_release = p
+
+    if args.efi_arch is None:
+        args.efi_arch = guess_efi_arch()
+
+    if args.stub is None:
+        args.stub = f'/usr/lib/systemd/boot/efi/linux{args.efi_arch}.efi.stub'
```
No, because we figure out `efi_arch` first. Can't do this step before.

behrmann replied:
The UI here went a bit awry. You are definitely right for `args.stub`, but `guess_efi_arch` could be the default for `args.efi_arch`, which would also show a user what arch would be guess if none were passed. The same is true for the path of the os-release.

Foxboron commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
```
Currrently `mkinitcpio` and `dracut` implement semantics around `/etc/kernel/cmdline` and `/etc/cmdline.d` which could be implemented in this helper instead. This would remove code duplication. The flag itself could take a singular file, but preferably default to reading `/etc/cmdline.d` is specified.

keszybz replied:
Hmm, but what is the point of those? If anything, I think those should be added to `kernel-install` first.

Foxboron replied:
I'm not sure what you are referring to. The initrd generators or the files themself?

keszybz replied:
Sorry for the lack of clarity. I meant that the support for those files when writing a boot loader entry. `kernel-install` looks at `/etc/kernel/cmdline` and for it is the primary tool for dealing with commandline entries. Thus, support for additional configuration files for the command line entry should be added there first.

Foxboron replied:
dracut has deprecated `/etc/kernerl/cmdline` for their UKIs. I'm not sure why `kernel-install` is the reference for what is and isn't supported?

Foxboron replied:
I'm confused and this was apparently more complicated?

Dracut itself reads from `/etc/cmdline` which is deprecated in favour of `/etc/cmdline.d/*.conf`. But their `kernel-install` hooks do read `/etc/kernel/cmdline`.

I don't understand why all these paths are needed.

poettering replied:
kernel-install is a systemd thing, and so is /etc/kernel/cmdline.

/etc/cmdline otoh might be something the dracut people came up with?

Foxboron replied:
I think so. I do wonder if it's better to standardize these paths for the sake of consistency?

dtardon replied:
`/etc/cmdline` and `/etc/cmdline.d/*.conf` aren't related to kernel-install. dracut uses these paths to allow to store additional kernel cmdline arguments in the initrd itself. When the initrd runs, dracut treats them as if they were part of the real cmdline. (This, obviously, only works if everything accesses the cmdline via dracut-provided functions. Which causes problems these days as some stuff is implemented as units/generators, that only see the cmdline systemd gives them...)

Foxboron commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
```
If we do want to hardcode `sbsign` usage then we need to be more clever here. `sbsign` supports `--engine` which allows you to specify an arbitrary string for `--key`. 

It would allow us to utilize PKCS11 to sign with a TPM and/or hardware keys.

Preferably I'd prefer if we leave this out of the helper, but that is personal and not very strong opinion.

keszybz replied:
I think it's nice to do signing, at least optionally. In particular, I think it's nice to call `systemd-measure` too, because we have all the sections specified here. And then it seems reasonable to have all the parts in the same tool.

I'll do the following: add `--sign-engine` and pass that through as `--engine` to sbsign, along with the key and cert args.

poettering replied:
so i think right from the beginning we should clarify in the option names what the key/cert is used for. i.e. ultimately we probably want at least three sets of keys involved:

1. one key for SB signing the UKI
2. one key for PCR signing the UKI for the initrd boot phase
3. one key for PCR signing the UKI for all boot phases from initrd until shutdown begins.

Yes, we need distinct keys for the latter: people must enroll their LUKS volumes or creds with one of the two keys depending on whether they want it to be accessible after initrd transition or only before.

keszybz replied:
I have `--pcr-private-key`, `--pcr-public-key`, `--sb-engine`, `--sb-key`, `--sb-cert`. ("sb" instead of "secure-boot" because this abbreviation should be pretty clear in this contents and the commandline is annoying long already.)

poettering replied:
i see no reason to shorten the options needlessly. This is not an interface people use daily, but a build tool. I think expressivity is a good thing here. i.e. immediatelly grokking what this command line does is more important than to be able to type it fast. Hence "secureboot" is a much more common syntax than "sb".

Also, "--pcr-private-key=" vs."--sb-key="? Why does one get the "private" and the other doesn't?

I'd always qualify with "private" given we are talking about key pairs, and we should make sure we never confuse things between the public and private of the two...

Also as mentioned I am pretty sure we should have two key pairs for the PCR case.

keszybz replied:
> * one key for PCR signing the UKI for the initrd boot phase
> 
> * one key for PCR signing the UKI for all boot phases from initrd until shutdown begins.

Hmm, but `systemd-measure sign` doesn't seem to support phases at all? I'm all confused now. Shouldn't the output consist of multiple combos of `pkfp` + `pol` + `sig`?

poettering replied:
hmm, somwhere my commit for measure.c got lost that added that. fuck. must be a rebase fuckup.

There was supposed to be a hookup in verb_sign() much like in verb_calculate() for measure_phase(). But apaprently that got lost. Yikes.

bluca replied:
is that something to fix for 252?

poettering replied:
#25202

poettering replied:
I prepped #25224 now, to allow generating multiple signatures by multiple different keys.

medhefgo commented:
> I think that this should be done as a step of the building of the kernel, if at all.

That might be true for UKIs built by distros. But right now these are still done locally. And just in case the secure boot override does not work on some firmware it should be signed by whatever key the UKI is being signed with…

bluca commented:
> > I think that this should be done as a step of the building of the kernel, if at all.
> 
> That might be true for UKIs built by distros. But right now these are still done locally. And just in case the secure boot override does not work on some firmware it should be signed by whatever key the UKI is being signed with…

With sbverity --list it's easy to check whether something is signed or not - if the kernel is already signed, then we should not add a signature I think, but maybe as a fallback it's fine

poettering commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+    args = p.parse_args()
+
+    if bool(args.key) ^ bool(args.cert):
+        raise ValueError('--key and --cert must be specified together')
```
`--key=` and `--cert=`?

poettering commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,187 @@
+                for name, content, offset in sections),
+        ]
+        print('+', shell_join(measure))
+        subprocess.check_call(measure)
```
hmm, the idea is to calculate the expected PCR values and *sign* them, before you link things together with objcopy.

i.e.

`systemd-measure sign` → `objcopy` → `sbsign`

keszybz replied:
Indeed. The code does that now.

poettering commented on src/kernel-install/ukify.py:
```diff
@@ -21,6 +21,16 @@
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
```
wouldn't it make more sense to have some dictionary for this mapping? that maps native arch, and reported fw_platform_size value and maps it to another arch? i.e. even if the map will only have a single entry it sounds nicer to separate the code that reads the platform size from the table that maps it.

poettering commented:
(I think it would be wise to make the tool look for signing keys automatically in /etc/kernel/ btw)

poettering commented:
btw, might make sense to start embedding the kernel's `uname -r` string in a `.uname` section right-away. We need this to make UKIs recognizable as such for shim, as well as for ordering multiple kernels associated with the same os-release in sd-boot. Even if they don't consume this right-away I think this is straight-forward enough to just add.

medhefgo commented:
I fail to see why shim would need a uname section for that. The osrel section is the canonical way to identify UKI images already (it's mandatory to have one even).

Also, we should stop making new section names that start with a dot. Afaik those are supposed to be reserved for the implementation (for whatever you consider implementation here). A simple `uname` section works just as well.

bluca commented:
> I fail to see why shim would need a uname section for that. The osrel section is the canonical way to identify UKI images already (it's mandatory to have one even).
> 
> Also, we should stop making new section names that start with a dot. Afaik those are supposed to be reserved for the implementation (for whatever you consider implementation here). A simple `uname` section works just as well.

It will be possible to have multiple UKIs to one root fs, so it would be useful to distinguish them - just like on debian or ubuntu you can install multiple kernel packages that boot the same root filesystem, and they get separate entries in the grub menu

poettering commented:
> I fail to see why shim would need a uname section for that. The osrel section is the canonical way to identify UKI images already (it's mandatory to have one even).

Well, sure, it could look for .kernel too. But I think ".uname" is vaguely nicer, since it very clearly says "Linux kernel" to me. ".kernel" otoh just indicates "some OS". And ".osrel" communicates more than just a kernel, it communicates a "whole OS". But of course, this is just nitpicking, either of the three is fine.

however, we need the .uname thing either way for sd-boot if we want to allow multiple kernels per OS version, like fedora does it now, because otherwise we couldn't distinguish them. hence, ignore the shim angle maybe, focus on the sd-boot angle: we need a way to distinguish multiple kernels of the same distro for the sd-boot menu.

> Also, we should stop making new section names that start with a dot. Afaik those are supposed to be reserved for the implementation (for whatever you consider implementation here). A simple uname section works just as well.

I don't they they are reserved. Quoting MS:

> Microsoft typically prefixes their section names with a period, but it's not a requirement. For years, the Borland linker used section names like CODE and DATA.

(https://learn.microsoft.com/en-us/archive/msdn-magazine/2002/february/inside-windows-win32-portable-executable-file-format-in-detail)

Other docs suggest the first char must be a dot (https://blog.kowalczyk.info/articles/pefileformat.html) so I think that claim is simply garbage and wrong.

So I think we should stick to our naming scheme for now.

medhefgo commented:
Also, I was not implying there is no use for a uname section. It just makes no sense to use it for shim to detect an all-in-one os loader. shim expects to load a boot loader, not an os. It *expects* it to participate in its protocol or aborts on boot services exit. The osrel section makes far more sense here.

poettering commented:
if we wanted to insert these sections into arbitrary binaries (like the we did for the package notes stuff, i.e. https://systemd.io/COREDUMP_PACKAGE_METADATA/) then I think what you point out would matter a lot, but here we have a much simpler case: we insert these sections into sd-stub, which is extremely focussed and pretty much all our code.

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,233 @@
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
```
I added a call to `PIL.Image.open()` to verify that the image looks valid. If we should have some check on dimensions or such, this can be easily extended.

keszybz commented:
> > > I think that this should be done as a step of the building of the kernel, if at all.
> > 
> > 
> > That might be true for UKIs built by distros. But right now these are still done locally. And just in case the secure boot override does not work on some firmware it should be signed by whatever key the UKI is being signed with…
> 
> With sbverity --list it's easy to check whether something is signed or not - if the kernel is already signed, then we should not add a signature I think, but maybe as a fallback it's fine

I added a call to `sbverify --list` and if that returns nothing, sign the image with `sbsign`.

keszybz commented:
> btw, might make sense to start embedding the kernel's `uname -r` string in a `.uname` section right-away. We need this to make UKIs recognizable as such for shim, as well as for ordering multiple kernels associated with the same os-release in sd-boot. Even if they don't consume this right-away I think this is straight-forward enough to just add.

I agree that it'd be nice to expose `uname -r`. Ideally we'd just extract that from the image, but that seems very hard. I'll a parameter for this now. For use cases like distro builds this would be enough, because the version number is readily available.

DaanDeMeyer commented:
Should we allow file based configuration as well for this? In mkosi, I find it very nice that we can just have a mkosi.repart/ directory to allow users to configure repart without having to add plumbing for all of repart's options in mkosi. It'd be great if we could do the same for this tool and other tools as well.

keszybz commented:
Update. I tried to incorporate all feedback, but since github doesn't make it very easy to respond to all comments, I might have missed something.
Bigger changes:
- `pefile` module is used to get the offsets
- section offsets are calculated as `offset-of-previous-section + round(size-of-previous-section, 4k)`
- `.linux` is now last
- if the input linux image is unsigned, it'll be signed with the secure-boot signing key
- optionally, signing of linux can be disabled or forced with `--sign-kernel`, `--no-sign-kernel`
- options names are not abbreviated: `--secureboot-engine`, `--secureboot-private-key`, `--secureboot-certificate`, `--pcr-private-key`, `--pcr-public-key`
- PCR banks can be selected with `--pcr-banks`
- splash image is loaded using pillow to check format correctness
- additional sections can be specified with `--section` (and file name or content string)
- `--tools` option can be used specify something like `--tools=build/` to use binaries from that directory
- `systemd-measure` is called first to generate a signed statement and this is included in the `.pcrsig` section
- `--uname` can be used to specify contents of `.uname` section.
- the program is installed as `/usr/lib/systemd/ukify`.

TODO:
- figure out if `--set-section-flags` shall be used (`objcopy` isn't happy, see my comment above)
- multiple signing keys for pcrs
- is the `.pcrkey` section useful?
- `.uname` should be read directly from the linux image. But the format doesn't seem to be standardized, so this is messy. If you know a clean way to do this, please speak up.
- more testing. I did only some superficial testing with booting of an amd64 VM.

keszybz commented:
> Should we allow file based configuration as well for this? In mkosi, I find it very nice that we can just have a mkosi.repart/ directory to allow users to configure repart without having to add plumbing for all of repart's options in mkosi. It'd be great if we could do the same for this tool and other tools as well.

I'm not sure yet if this will be useful. Arguments like linux and initrd are generally volatile and would normally be passed on the commandline. And Lennart suggested loading keys from `/etc/kernel`. The only thing that would be specified in "normal" use would be the commandline, and I'm not sure if it's worth having a configuration file for that. (The commadnline can be specified as `--cmdline=@file` already.)

bluca commented on meson_options.txt:
```diff
@@ -501,6 +501,8 @@ option('llvm-fuzz', type : 'boolean', value : false,
        description : 'build against LLVM libFuzzer')
 option('kernel-install', type: 'boolean', value: true,
        description : 'install kernel-install and associated files')
+option('ukify', type: 'boolean', value: true,
```
any reason this needs an option? it's just a script without any dependency or side effects, no?

keszybz replied:
The same is true for `kernel-install`. I expect that some folks will be unhappy that it's in python, so I added the option to match surrounding code.

bluca commented:
Excellent stuff - too soon for a manpage? What about a test case?

keszybz commented:
> Excellent stuff - too soon for a manpage? What about a test case?

Excellent question ;)

Let's wait with the man page until the interface settles down. I added descriptions to all options, so `--help` output is fairly useful by itself. It's possible that if we add multiple signing keys the options interface might morph quite a bit.

And as for tests: do we even test systemd-boot anywhere?

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,465 @@
+
+    p.add_argument('--section',
+                   dest='sections',
+                   metavar='NAME:TEXT|@PATH[:OFFSET]',
```
I forgot to update this (`:OFFSET` is now gone).

bluca commented:
> And as for tests: do we even test systemd-boot anywhere?

It's tested implicitly by mkosi - it's coarse in the sense that it either boots or doesn't.

But a simple build test, to make sure the script doesn't error out unexpectedly and such, would be a good starting point - with a simple sanity check that verifies all the expected sections are in the PE for example

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,465 @@
+    except ImportError:
+        return
+
+    img = Image.open(args.splash)
```
I'll leave that as TODO.

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,465 @@
+                   help='embedded public key to seal secrets to [.pcrpkey section]')
+    p.add_argument('--uname',
+                   metavar='VERSION',
+                   help='"uname -rv" information [.uname section]')
```
I was thinking that somebody might build multiple kernel in development, and than have `#1 …`, `#2 …` in the version part. But indeed, that's a fringe case. I'll drop the `v`.

poettering commented:
> * is the `.pcrkey` section useful?

Yes, I think so. It's the "default" key that userspace should lock its stuff to, i.e. the key to bind to if you don't know better what to bind again. i.e. the key that covers all relevant userspace boot phases where secrets might be unlocked, everything from "enter-initrd" until "enter-initrd:leave-initrd:sysinit:ready" really.

So, it's the "riskless" key: during regular userspace runtime this is what you can lock your stuff against and it will work.

And yes, I would expect that in future we'll sign some boot phases with other keys, but that would be for the more finegrained cases, for example, for allowing to unlock the root fs just in the initrd. But I'd claim that is a singular, special case, and thus should be the exception, not the default.

Hence, yes, I absolutely think .pcrpkey makes sense.

poettering commented:
> * `.uname` should be read directly from the linux image. But the format doesn't seem to be standardized, so this is messy. If you know a clean way to do this, please speak up.

everybody (kmod, kernel-install, …) use `uname -r`, so should we here. it's as standardized as it could get and shows up in the file system all over the place, for example in `/lib/modules/…`.

poettering commented:
> And as for tests: do we even test systemd-boot anywhere?

no, we don't. and we should change that.

bluca commented:
> > And as for tests: do we even test systemd-boot anywhere?
> 
> no, we don't. and we should change that.

There's no detailed feature-level checks, but it is used to boot mkosi-built images, so basic sanity checks are done implicitly

DaanDeMeyer commented:
(Testing systemd-boot becomes very easy once we switch to using mkosi for the integration tests)

medhefgo commented:
It's probably best to have this convert the splash image itself (also, maybe this should be using pillow?).

I converted the arch logo form the website to bmp with `convert archlinux-logo-dark-1200dpi.b42bd35d5916.png -resize 800 arch.bmp` and the image renders colors horribly in qemu.

bluca commented:
Is this supposed to work for foreign architectures? Trying to build an arm64 from x86_64 with binutils 2.39:

```
$ ./ukify.py --tools /usr/bin/ --cmdline "" --os-release @uki/osrelease --stub uki/linuxaa64.efi.stub --efi-arch aa64 uki/linux uki/initrd 
+ /usr/bin/objcopy uki/linuxaa64.efi.stub --add-section .osrel=uki/osrelease --change-section-vma .osrel=0x11000 --add-section .initrd=uki/initrd --change-section-vma .initrd=0x12000 --add-section .linux=uki/linux --change-section-vma .linux=0x2478000 linux.unsigned.efi
/usr/bin/objcopy: uki/linuxaa64.efi.stub: file format not recognized
/usr/bin/objcopy: --change-section-vma .linux=0x0000000002478000 never used
/usr/bin/objcopy: --change-section-vma .initrd=0x0000000000012000 never used
/usr/bin/objcopy: --change-section-vma .osrel=0x0000000000011000 never used
Traceback (most recent call last):
  File "/tmp/./ukify.py", line 471, in <module>
    main()
  File "/tmp/./ukify.py", line 467, in main
    make_uki(args)
  File "/tmp/./ukify.py", line 306, in make_uki
    subprocess.check_call(cmd)
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '[PosixPath('/usr/bin/objcopy'), PosixPath('uki/linuxaa64.efi.stub'), '--add-section', '.osrel=uki/osrelease', '--change-section-vma', '.osrel=0x11000', '--add-section', '.initrd=uki/initrd', '--change-section-vma', '.initrd=0x12000', '--add-section', '.linux=uki/linux', '--change-section-vma', '.linux=0x2478000', 'linux.unsigned.efi']' returned non-zero exit status 1.
$ file uki/linuxaa64.efi.stub 
uki/linuxaa64.efi.stub: MS-DOS executable PE32+ executable (EFI application) Aarch64 (stripped to external PDB), for MS Windows
```

medhefgo commented:
> Is this supposed to work for foreign architectures? Trying to build an arm64 from x86_64 with binutils 2.39:

binutils does not do foreign architectures. You'll need to use llvm-objcopy or the one from the appropriate aarch64 binutils package. Why the gnu toolchain is not multi target is something I'll probably never understand…

keszybz commented:
> > * `.uname` should be read directly from the linux image. But the format doesn't seem to be standardized, so this is messy. If you know a clean way to do this, please speak up.
> 
> everybody (kmod, kernel-install, …) use `uname -r`, so should we here. it's as standardized as it could get and shows up in the file system all over the place, for example in `/lib/modules/…`.

The problem is that **extracting** of the `uname -r` value from an image is hard. You can guess from the file name, maybe, but it would be much much nicer to read the file contents. This is the part I don't know how to do.

keszybz commented on src/kernel-install/ukify.py:
```diff
@@ -0,0 +1,472 @@
+        ('dtb',     args.devicetree, True ),
+        ('splash',  args.splash,     True ),
+        ('pcrpkey', pcrpkey,         True ),
+        ('initrd',  args.initrd,     True ),
```
Right now we use the file multiple times: possibly twice for measurement and once as input for `objcopy`. We could save it to a tempfile if the input is a pipe. We could also accept multiple `--initrd` params and concatenate them. We could also do both… I think is is more of an optional feature, so I'm leaving it for later. I'll add it to the TODO in comment.

behrmann replied:
@medhefgo Maybe a bit too hacky, but `@<(cat initrd1 initrd2) should parse as the correct input, since it will passed in as the string `@/dev/fd/12345` and `/dev/fd/12345` will then end up in `args.initrd`. This might still break in other places, I only checked the initial argparse bit.

keszybz commented:
Updated. I added a bunch of tests, but those depend on #25401 for JSON output. So please ignore patches before `meson: resort imports` here, and instead comment in #25401 if appropriate.

I moved the file to a new subdirectory, because there are tests and a bunch of extra files and `src/kernel-install` would become cluttered. After adding the tests I found a bunch of small bugs, and those are of course now fixed. If  you like the general approach, I'll work on the man page now.

bluca commented:
there will be (extremely dumb) repository scanning tools that will go ham when they see a private key file called "src/ukify/signing.key" checked in the repo - maybe rename that to "test.key" so that it is immediately obvious to said dumb tools what's it for? maybe also move the test files and case under a ukify/test/ subdir?

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -69,6 +70,118 @@ def round_to_page(x, page_size=4096):
+
+    @classmethod
+    def scrape_x86(cls, filename, opts=None):
+        # Based on https://gitlab.archlinux.org/archlinux/mkinitcpio/mkinitcpio/-/blob/master/functions#L136
```
I think it's better to link to official kernel docs for this (looking for the `kernel_version` offset. The description for it also contains some further rules for it): https://www.kernel.org/doc/html/latest/x86/boot.html#the-real-mode-kernel-header

medhefgo commented:
Why not generate a test key as part of the build (or meson setup)?

Also, at least on debian based distros you can actually use `/usr/share/ovmf/PkKek-1-snakeoil.key` (password `snakeoil`). It comes with a snakeoil ovmf build that already trusts this key for secure boot.

Foxboron commented:
@keszybz Happy you found the `mkinitcpio` implementation useful :)

behrmann commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+        # python3-lzo is not packaged for Fedora
+        raise NotImplementedError('lzo decompression not implemented')
+
+    if start.startswith(b'BZh'):
+        raise NotImplementedError('bzip2 decompression not implemented')
+
+    if start.startswith(b'\x5d\x00\x00'):
+        raise NotImplementedError('lzma decompression not implemented')
```
Why not?
```suggestion
    if start.startswith(b'BZh'):
        import bz2
        return bz2.open(f).read()

    if start.startswith(b'\x5d\x00\x00'):
        import lzma
        return lzma.open(f).read()
```
They're already in the stdlib.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+class Section:
+    name: str
+    content: pathlib.Path
+    tmpfile: typing.Optional[typing.IO] = None
+    offset: typing.Optional[int] = None
```
Since you seem to be using this below already, since the introduction of the type union operator in 3.10, most people seem to favour 
```suggestion
    tmpfile: typing.IO | None = None
    offset: int | None = None
```
instead of `Optional`, since it is a bit more concise and readable.

keszybz replied:
I like this. 3.10 is present in Debian testing and all releases of Fedora, so it should be OK. If people want to support earlier, we can always mutilate the code in a later patch.

behrmann commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+
+@dataclasses.dataclass
+class UKI:
+    executable: [pathlib.Path|str]
```
Should this be this?
```suggestion
    executable: list[pathlib.Path|str]
```

keszybz commented:
> Why not generate a test key as part of the build (or meson setup)?

It'd be possible, but it's yet-another thing to do during tests. The goal of those tests is not to test openssl, but to test our tools. It also makes the tests a bit faster, since there's not need to generate random numbers and such.

> Also, at least on debian based distros you can actually use `/usr/share/ovmf/PkKek-1-snakeoil.key` (password `snakeoil`). It comes with a snakeoil ovmf build that already trusts this key for secure boot.

I'd prefer to avoid integration with distro-specific stuff. We can always add that later.

DaanDeMeyer commented:
Should this have --root and --image options?

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+        # host_arch glob : [efi_arch, 32_bit_efi_arch?]
+        'x86_64'       : ['x64', 'ia32'],
+        'i[3456]86'    : ['ia32'],
+        'aarch64'      : ['aa64', 'arm'],
```
Only x86 has mixed mode so far, this having a fallback here makes no sense.

keszybz replied:
Dropped.

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+
+    # import is delayed, to avoid import when the splash image is not used
+    try:
+        from PIL import Image
```
Isn't PIL python2? Afaik, Pillow is the go-to now.

behrmann replied:
The fork Pillow took over `PIL` as module, the same import is [used in their tutorial](https://pillow.readthedocs.io/en/stable/handbook/tutorial.html).

keszybz replied:
Yeah, `import PIL` means `pillow` for the last few years.

medhefgo replied:
I knew there is a reason why I don't do python…

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,592 @@
+    except ImportError:
+        return
+
+    img = Image.open(filename)
```
This should really pass the format arg to restrict it to bmp.

medhefgo reviewed (commented):
I gave this a spin and seems to be working like a charm now. The assembled stub now works without the `-O1` workaround, so I would really like to see mkosi to make use of this so that it can go away.

poettering commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,476 @@
+        ('splash',  opts.splash,     True ),
+        ('pcrpkey', pcrpkey,         True ),
+        ('initrd',  opts.initrd,     True ),
+        ('uname',   opts.uname,      False),
```
this should be measured, really.

poettering replied:
(i mean, a patch adding measurement of this to sd-stub is trivial, might include in some PR)

keszybz replied:
OK. But let's leave it for later…

poettering commented on src/ukify/ukify.py:
```diff
@@ -219,6 +332,9 @@ def make_uki(opts):
         linux = opts.linux
 
+    if opts.uname is None:
+        opts.uname = Uname.scrape(opts.linux, opts=opts)
```
grmbl, do we really want this? isn#t this better something we expect people to pass in? i mean, during build time this information should be clearly available always?

i mean, this info isn#t strictly necessary, so for people who quickndirty want to build a UKI they can just leave the field out?

poettering replied:
oh, hmm, kernels are internally ELF, so i guess we could use the package notes infra as primary source of info, no? i mean, it would be a form of dogfooding ;-)

keszybz replied:
I don't like this part too much, it is hacky. And in stuff like package installation scripts it won't be necessary, since you generally know the uname from the package version. But for casual use where people combine an existing kernel with some commandline options and other fluff, I think it's fairly useful. If it turns out that this scraping is too brittle, we can always drop that later.

>  we could use the package notes

I don't think anybody builds the kernel with package notes. On PPC64, the kernel has a note with the version, but afaict, it's not on all architectures and is hard to access anyway.

poettering replied:
maybe at the very least add a log message "Kernel version was not speicified, resorting to scraping. Yuck. Please specify explicitly." or something like that.

plus an emoji.

keszybz replied:
I used "confounded face": 😖

poettering reviewed (approved):
looks excellent (well, the json changes from the other PR i am not so convinced about, see other PR)

poettering commented:
for x86 kernel the arch is in boot_params structure btw, i.e. you can directly parse it from the beginning of the image...

keszybz commented:
> for x86 kernel the arch is in boot_params structure btw, i.e. you can directly parse it from the beginning of the image...

This is already done, see `scrape_x86()`.

--

Updated. Changes:
- the file is now processed with jinja2 so that --version works
- signing with multiple keys is now supported
- man page
- there's a bunch of additional tests for the signing
- bugfixes and other small changes based on @medhefgo's comments

I squashed some pull requests because I made too many fixups and it didnt' make sense to keep the old split.
I'm still relying on the having conforming JSON output in the tests. Thus, all commits except the last five should be ignored. Let's continue the discussion about earlier commits in #25401.

keszybz commented on src/ukify/test/test_ukify.py:
```diff
@@ -0,0 +1,368 @@
+    ],
+        text=True)
+
+    assert open(tmpdir / 'out.pcrpkey').read() == pub.read_text()
```
Oh, just punish me harder.

keszybz commented:
The fight with CI begins.

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,700 @@
+    except ImportError:
+        return
+
+    img = Image.open(filename, formats=['BMP'])
+    print(f'Splash image {filename} is {img.width}×{img.height} pixels')
```
I just realized, you can just do this to convert to bmp and allow any image format:
```
img.save(…, format='BMP')
```

keszybz replied:
Thansk, this looks pretty nice. I'll leave that for later though… I didn't actually test the splash functionality at all. We probably should add some real test for it before building conversion functionality.

medhefgo replied:
I tried it and the added splash is working (even converted ones from pillow). And it will have correct colors once https://github.com/systemd/systemd/pull/25498 lands.

medhefgo commented on meson_options.txt:
```diff
@@ -501,6 +501,8 @@ option('llvm-fuzz', type : 'boolean', value : false,
        description : 'build against LLVM libFuzzer')
 option('kernel-install', type: 'boolean', value: true,
        description : 'install kernel-install and associated files')
+option('ukify', type: 'boolean', value: true,
```
Can we please start making new meson options use actual meson feature type instead of bools?

keszybz replied:
I think we'll want to convert all the options to use features (and automatic feature selection) in one fell swoop. I don't think it makes sense to use a mixed approach.

keszybz replied:
I changed the typo to combo 'true'/'false'/'auto' now. The code doesn't work with python3.9 so it's only enabled for >= 3.10.

medhefgo commented on man/ukify.xml:
```diff
@@ -0,0 +1,294 @@
+      <varlistentry>
+        <term><option>--devicetree=<replaceable>PATH</replaceable></option></term>
+
+        <para>Specify the devicetree description (the <literal>.dtb</literal> section). The argument is a
+        path to a DeviceTree file. If not specified, the section will not be present.</para>
```
This should really say that it's a compiled binary dt…

medhefgo commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,571 @@
+        return self.content.stat().st_size
+
+    def check_name(self):
+        # PE section names with more than 8 characters are legal, but our stub does
+        # not support them. A dot is prepended, so the name must be at most 7 chars.
+        name = f'.{self.name}'
```
I don't think this should force section names to start with a dot. It seems too restrictive to me and seems inconsistent as other places/tools you'd have to pass the dot…

Also a light check for duplicates would nice as objcopy create a really unhelpful error (`can't add section '.cmdline': file in wrong format`).

keszybz replied:
Both done.

medhefgo commented:
Also, would be nice if you could pass more than one initrd and have ukify automatically concat them. (Useful for cpu microcode)

keszybz commented:
Updated:
- section names without a dot
- multiple initrds
- more workarounds for CI

keszybz commented:
Code scanning results / CodeQL — stupid warnings about `assert` in tests
ci/semaphoreci/pr: Debian autopkgtest (LXC) — "network is unreachable"

keszybz commented:
Updated. I dropped the changes to userdbctl, but left in the changes for bootctl. Bootctl has very short output so streaming output is not important there. I hope we can get this merged now. There's still a bunch of features that can be added, incl. the list in one of the comments above, and more tests, etc., but I think it's fine to add it later in individual PRs.

keszybz commented:
CentOS CI (Arch Linux + sanitizers) — TEST-29-PORTABLE
jammy-ppc64el — TEST-29-PORTABLE:                   FAIL     (1210 s)
others — network issues with apt

poettering commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,573 @@
+
+    uki = UKI(opts.stub)
+
+    # TODO: derive public key from from opts.pcr_private_keys?
```
i think we should do that in systemd-measure automatically btw

bluca commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,573 @@
+
+    p.add_argument('--secureboot-engine',
+                   dest='sb_engine',
+                   help='OpenSSL engine to use for signing')
```
this is not actually passed to sbsign?

bluca commented on src/ukify/ukify.py:
```diff
@@ -0,0 +1,573 @@
+                   metavar='BANK…',
+                   type=parse_banks)
+
+    p.add_argument('--secureboot-engine',
```
We need to be able to use an engine with systemd-measure too - it doesn't right now, but it will need for production usage. Hence, probably better to just call this "signing-engine" or "openssl-engine" or so, so that it can just be used for measure in the future without causing confusion or requiring param name change

poettering replied:
but you might want to use different engines for both?

it appears to me the engine options should be called like the key options they belong to.

keszybz replied:
I didn't use "openssl" in the name on purpose, because whether OpenSSL is used by sbsign is an internal implementation detail. I expect that we'll grow other mechanisms in the future. OTOH, I think "secureboot" is appropriate also for the case where this would be passed to systemd-measure. It'd a secureboot signature that would be created.

bluca replied:
mmh actually you can use the tpm2 measurements without secure boot? That's what the SUSE folks do, no?

bluca replied:
> but you might want to use different engines for both?

That seems very unlikely to me. Different keys yes, but I can't think of a build system that would use different signing engines during the same build.

bluca commented on src/ukify/test/test_ukify.py:
```diff
@@ -0,0 +1,368 @@
+
+    items = json.loads(text)
+
+    for item in items:
```
still needs to be handled

keszybz replied:
It was handled by closing the alert. I really don't think there's anything to fix here. The code looks better with the `else`.

bluca commented on src/ukify/test/test_ukify.py:
```diff
@@ -0,0 +1,380 @@
+
+try:
+    # pyflakes: noqa
+    import pefile  # noqa
```
still needs to be handled

keszybz replied:
No, it's a false positive.

poettering commented on src/ukify/ukify.py:
```diff
@@ -453,6 +454,14 @@ def make_uki(opts):
+        case [initrd]:
+            pass
+        case multiple:
+            initrd = b''.join(file.open('rb').read() for file in multiple)
```
you must pad each initrd to multiples of 32bit, with NUL bytes. (except for the last one)

poettering commented:
lgtm. but i am not a python-guru. the 32bit padding issue matters though

keszybz commented on man/ukify.xml:
```diff
@@ -0,0 +1,309 @@
+    <option>--pcr-private-key=</option>, <option>--pcr-public-key=</option>, and <option>--phases=</option>
+    arguments more than once. Signatures will be then performed with each of the specified keys. When both
+    <option>--phases=</option> and <option>--pcr-private-key=</option> are used, they must be specified the
+    same number of times, and then n-th boot phase path set will be signed by the n-th key. This can be used
```
I think "n-th" is easier to pronounce. "k-th key" is a bit of a mouthful.

"then" seems OK to me. You specify the keys, and then (later) something is done with them.

behrmann replied:
I think "then" there is ok, too, but then it's still missing an article, I think:
```suggestion
    same number of times, and then the n-th boot phase path set will be signed by the n-th key. This can be used
```

keszybz commented on src/ukify/test/example.tpm2-pcr-private2.pem:
```diff
@@ -0,0 +1,28 @@
+ou0aw7F6mZ2/M5G/zOm0qCbsWtaMSOSuA/6iU9pt40ZYAN4U/wfqnw2VEhFp7w1M
+zYZbBD0qeYdp4fsSnYqLfbAVlP/0zva3vL02REknVjPUJp/hjJUhA+mV7nvT6yV4
+q88Ieo8LwCW5sbmwirSpmw==
+-----END PRIVATE KEY-----
```
I've already received a bunch of reports from automatic scanners maintained by RH. I was actually impressed that they _immediately_ flagged something pushed to a branch. As a result of that, I renamed the files to `example.*`. I added base64 obfuscation now too.

bluca replied:
I am pretty sure more low-quality scanners will happily ignore the "test" and "example" prefixes and just barf anyway... Generating on the fly with openssl is not too difficult, any reason not to just do that?

keszybz replied:
openssl command wasn't used at all, so that'd pull that in, and require some code to generate the file. And the generation of the file would probably take longer than all the other tests combined, because openssl would gather "true" entropy.

keszybz commented:
Thank you all. Updated with the comments addressed.

poettering commented on man/ukify.xml:
```diff
@@ -78,8 +78,10 @@
-    arguments are mandatory.</para>
+    <para>Note that the <replaceable>LINUX</replaceable> positional argument is mandatory. The
+    <replaceable>INITRD</replaceable> positional arguments are optional. If more than one is specified, they
+    will all be combined into a signle section. This is useful to for example prepend microcode before the
```
signle → single

(section → PE section)

poettering commented:
lgtm, setting green label (not a total python guru myself, but looks good to me, and since luca liked it, i guess we are good)

keszybz commented on src/ukify/ukify.py:
```diff
@@ -404,6 +405,24 @@ def call_systemd_measure(uki, linux, opts):
+            for file in multiple:
+                initrd = file.read_bytes()
+                padding = b'\0' * round_up(len(initrd), 4)  # pad to 32 bit alignment
+                seq += [initrd, padding]
```
Yeah, it makes the code simpler and doesn't make any practical difference, because there's always a following section (`.linux`)  and it's aligned anyway.

keszybz commented:
Updated with the typo fix.

yuwata commented:
The discussion about the json format was finished?

keszybz commented:
> The discussion about the json format was finished?

It was sidestepped. The patches for userdbctl were dropped from this PR. They are still in #25401, which I plan to return at a later point.

yuwata commented:
But this still contains the one for bootctl. Is it intended?

poettering commented:
yeah, there was a discussion about that on the video call the other week. bootctl shouldn't ever deal with 10k boot entries, but user databases might. hence doing the full table in bootctl is probably OK.
